### PR TITLE
23 train classifiers

### LINKED ISSUE #23: Train classifiers
Follow a progression from simplest to more complicated. Something along the lines of:

1. Logistic regression (multioutput)
2. XGBoost
3. Fine-tune the embeddings

### PR DESCRIPTION
**Pull request by RFOxbury**
---

# Description

Changes on this PR:

- Some notebooks moved location - ignore these changes
- The metrics used to evaluate the baseline models have been changed (the process of training classifiers prompted more thinking about what metrics would be useful)
- There is code to train basic classifiers
- There is some WIP code towards fine-tuning a transformer model

Fixes #23 

# Instructions for Reviewer

To review this PR:

- Run `pipeline/models/baseline_model.py` and try varying the model type ("most_probable"/"majority_combination") and the wandb input (True to log a run on wandb, False to run the model without logging a run)

To check the classifiers, either
- Run the notebook `05_train_classifiers.py`
- AND/OR run `pipeline/models/classifiers.py` (a refactored version of the notebook above) with both `wandb=False` and then `wandb=True`.

The notebook and the script contain largely the same code but the heatmap table works better in the notebook, and the notebook explains a bit more of the thinking (eg arriving at the decision to use macro averaging).

Please pay special attention to ...

# Checklist:

- [x] I have refactored my code out from `notebooks/`
- [x] I have checked the code runs
- [x] I have tested the code
- [x] I have run `pre-commit` and addressed any issues not automatically fixed
- [x] I have merged any new changes from `dev`
- [x] I have documented the code
  - [x] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [x] I have explained this PR above
- [x] I have requested a code review

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1,20 +1,22 @@
 altair
-hdbscan==0.8.33
-metaflow==2.9.14
-numpy==1.23.4
+hdbscan
```
Comment (j-gillam) [inline requirements.txt:2]: Why did you remove the package versions? Was it causing dependency issues?

File: discovery_child_development/utils/classification_utils.py [hunk 2]
```diff
@@ -95,3 +130,164 @@ def flatten_classification_report(report):
...
+    return df.reset_index()
+
+
+# Example usage:
+# label_name = 'ar / vr'
+# false_negatives_identifiers = get_false_negatives_identifiers(Y_val, predictions_val, label_name)
+# print(false_negatives_identifiers)
+
+
+def categorise_predictions(label, predictions, actual):
```
Comment (j-gillam) [inline discovery_child_development/utils/classification_utils.py]: Could you add a doc string to explain this function? (All the ones above are great, super readable!)

File: discovery_child_development/pipeline/models/classifiers.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+from discovery_child_development.pipeline.models import baseline_model as bm
+from discovery_child_development.utils import classification_utils
+from discovery_child_development.utils import cluster_analysis_utils as cau
+from discovery_child_development.utils import wandb as wb
+from discovery_child_development.utils.io import import_config
+
+load_dotenv()
+
+S3_BUCKET = os.environ.get("S3_BUCKET")
+PARAMS = import_config("config.yaml")
```
Comment (j-gillam) [inline discovery_child_development/pipeline/models/classifiers.py]: (Again feel free to ignore). But I noticed that you created an import_config function. One actually already exists (which I only found about like 3 months ago ahah) in the __init__ file inside the discovery_child_development folder. So you can use the way its set out there instead (if you wanted to). I have used it the stuff I've done so far, if you were interested.

File: discovery_child_development/pipeline/models/classifiers.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+from discovery_child_development.getters import openalex as oa
+from discovery_child_development.pipeline.models import baseline_model as bm
+from discovery_child_development.utils import classification_utils
+from discovery_child_development.utils import cluster_analysis_utils as cau
+from discovery_child_development.utils import wandb as wb
+from discovery_child_development.utils.io import import_config
+
+load_dotenv()
+
+S3_BUCKET = os.environ.get("S3_BUCKET")
```
Comment (j-gillam) [inline discovery_child_development/pipeline/models/classifiers.py]: (Feel free to ignore this as it doesn't change functionality at all) The S3 bucket is actually already defined in the __init__ file. It is called BUCKET_NAME instead of S3_BUCKET, but that is a quick fix. So if you wanted to you could use that instead of the load_dotenv stuff? As you can write something like:
`from discovery_child_development import S3_BUCKET`
In the PR I'll make in the next few days I've got an example and I also changed BUCKET_NAME to S3_BUCKET.

File: discovery_child_development/utils/classification_utils.py [hunk 5]
```diff
@@ -2,8 +2,16 @@
 Utils functions for training a classifier
```
Comment (j-gillam) [inline discovery_child_development/utils/classification_utils.py:2]: This file is brilliant! It was on my list of future things to think about creating, so fab you have already done it!

File: discovery_child_development/pipeline/models/baseline_model.py [hunk 6]
```diff
@@ -3,7 +3,7 @@
 
 Usage:
 
-python discovery_child_development/analysis/baseline_model.py --model_type most_probable --wandb True
+python discovery_child_development/pipeline/models/baseline_model.py --model_type most_probable --wandb True
 
 model_type can either be 'most_probable' or 'majority_combination' and determines the type of baseline classifier used.
```
Comment (j-gillam) [inline discovery_child_development/pipeline/models/baseline_model.py]: May you explain a bit more (maybe just a line or two) about the different between the two options, e.g. most probable vs majority combination?

File: discovery_child_development/pipeline/models/baseline_model.py [hunk 7]
```diff
@@ -3,7 +3,7 @@
 
 Usage:
```
Comment (j-gillam) [inline discovery_child_development/pipeline/models/baseline_model.py:4]: Excellent w&b use! I am for sure going to be staring at/stealing this for the binary classifier! 

Any chance you could add a doc string to the run_baseline_model? It is quite a big function so it was a little difficult to visualise what was happening at each part.

File: discovery_child_development/pipeline/models/classifiers.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+        # This is the required input to the sklearn MultiLabelBinarizer.
+        openalex_data.groupby(["openalex_id", "text"])["sub_category"]
+        .agg(lambda x: tuple(set(x)))
+        .reset_index()
+    )
+    # Set the index - useful later for creating training/validation split
+    openalex_data_wide = openalex_data_wide.set_index("openalex_id")
+
+    # Load embeddings
+    embeddings = S3.download_obj(
```
Comment (j-gillam) [inline discovery_child_development/pipeline/models/classifiers.py]: Would it be possible to refactor these out into getters? I think the cookiecutter advice tends to err on any time you load something in to have it in a getter (as you may want to load it somewhere else at some point). Not super important! More of a nice to have.

**Review by j-gillam — APPROVED** [review]
I am currently in the process of running the py file so will update if anything errors out but it all looks great! I have put in some comments, but they are super tiny so feel free to pick and choose if you want to action any, as none of them affect the output. 

I haven't inspected the notebooks too closely as I know they are more for testing.

File: requirements.txt (hunk 1 above)
Comment (RFOxbury) [reply requirements.txt:2]: Yes... I have already forgotten exactly what errors but yes 🙈 I removed all the package versions (except for the bigquery one, because I didn't want to tamper with that) and initialised the environment with python 3.10

File: discovery_child_development/utils/classification_utils.py (hunk 2 above)
Comment (RFOxbury) [reply discovery_child_development/utils/classification_utils.py]: Thanks Jess! Had to remember what I was trying to do with this one 😅

File: discovery_child_development/pipeline/models/classifiers.py (hunk 3 above)
Comment (RFOxbury) [reply discovery_child_development/pipeline/models/classifiers.py]: 🤯 can't believe I didn't find this out sooner! Super useful to know - thank you!

File: discovery_child_development/pipeline/models/classifiers.py (hunk 4 above)
Comment (RFOxbury) [reply discovery_child_development/pipeline/models/classifiers.py]: Same as above! Thank you!!

File: discovery_child_development/pipeline/models/classifiers.py (hunk 8 above)
Comment (RFOxbury) [reply discovery_child_development/pipeline/models/classifiers.py]: Thanks, good spot!
